File: backend-springboot/turing/src/main/java/com/ashelabs/turing/config/MqttConfig.java
package com.ashelabs.turing.config;

import lombok.Data;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.context.annotation.Configuration;
import org.springframework.validation.annotation.Validated;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.Max;

@Configuration
@ConfigurationProperties(prefix = "mqtt")
@Data
@Validated
public class MqttConfig {

    @NotBlank(message = "MQTT broker URL is required")
    private String brokerUrl;

    @NotNull
    @Min(1)
    @Max(65535)
    private Integer brokerPort = 1883;

    @NotBlank(message = "MQTT username is required")
    private String username;

    @NotBlank(message = "MQTT password is required") 
    private String password;

    @NotBlank(message = "MQTT client ID is required")
    private String clientId = "turing-server";

    @NotNull
    @Min(10)
    private Integer keepAliveInterval = 60;

    @NotNull
    @Min(1)
    private Integer connectionTimeout = 30;

    @NotNull
    private Boolean cleanSession = true;

    @NotNull
    @Min(1)
    @Max(10)
    private Integer reconnectAttempts = 3;

    @NotNull
    @Min(1)
    private Integer reconnectDelay = 5000;

    // MQTT Topic Configuration
    @NotBlank(message = "Base MQTT topic is required")
    private String baseTopic = "mitsubishi2mqtt";

    public String getFullBrokerUrl() {
        return String.format("tcp://%s:%d", brokerUrl, brokerPort);
    }

    public String getStateTopicPattern() {
        return baseTopic + "/+/state";
    }

    public String getSettingsTopicPattern() {
        return baseTopic + "/+/settings";
    }

    public String getCommandTopic(String roomId, String command) {
        return String.format("%s/%s/%s/set", baseTopic, roomId, command);
    }
}

File: backend/src/common/config/mqtt.config.ts
import { registerAs } from '@nestjs/config';

export default registerAs('mqtt', () => ({
  brokerUrl:
    process.env.MQTT_BROKER_URL ||
    `mqtt://${process.env.MQTT_BROKER_HOST || process.env.MQTT_BROKER || '10.0.16.3'}:${process.env.MQTT_BROKER_PORT || '1883'}`,
  broker:
    process.env.MQTT_BROKER ||
    `mqtt://${process.env.MQTT_BROKER_HOST || '10.0.16.3'}:${process.env.MQTT_BROKER_PORT || '1883'}`,
  username:
    process.env.MQTT_BROKER_USERNAME ||
    process.env.MQTT_USERNAME ||
    'mqtt-service',
  password:
    process.env.MQTT_BROKER_PASSWORD ||
    process.env.MQTT_PASSWORD ||
    'asus1122334455',
  clientId: process.env.MQTT_CLIENT_ID || `turing-server-${Date.now()}`,

  // MQTT connection settings
  keepAliveInterval: parseInt(process.env.MQTT_KEEP_ALIVE || '30', 10),
  connectionTimeout: parseInt(process.env.MQTT_CONNECTION_TIMEOUT || '60', 10),
  cleanSession: process.env.MQTT_CLEAN_SESSION === 'true',
  reconnectAttempts: parseInt(process.env.MQTT_RECONNECT_ATTEMPTS || '10', 10),
  reconnectDelay: parseInt(process.env.MQTT_RECONNECT_DELAY || '3000', 10),
  baseTopic: process.env.MQTT_BASE_TOPIC || 'mitsubishi2mqtt',
}));
